pkg/authz/responsefilterer.go

```diff
@@ -0,0 +1,745 @@
+type ResponseFilterer interface {
+	// RunPreFilters runs the filtering on the response, updating its body and status code
+	// based on the rules and authz data.
+	FilterResp(resp *http.Response) error
+}
+
+// WithResponseFilterer returns a copy of parent in which the responseFilterer value is set
+func WithResponseFilterer(parent context.Context, info ResponseFilterer) context.Context {
+	return context.WithValue(parent, responseFiltererKey, info)
+}
+
+// ResponseFiltererFrom returns the value of the responseFilterer key on the ctx
+func ResponseFiltererFrom(ctx context.Context) (ResponseFilterer, bool) {
+	return responseFiltererKey.Value(ctx)
+}
+
+// NewEmptyResponseFilterer creates a new ResponseFilterer that does not run any pre-filters.
+func NewEmptyResponseFilterer(restMapper meta.RESTMapper, input *rules.ResolveInput) ResponseFilterer {
+	rf := &StandardResponseFilterer{
+		restMapper: restMapper,
+		input:      input,
+
+		filteredRules: nil,
+		client:        nil,
+
+		prefilteredStarted: true,
+		preFilterCompleted: make(chan prefilterResult, 1),
+	}
+	rf.preFilterCompleted <- prefilterResult{allAllowed: true} // signal that pre-filters are completed, as this is an empty call
+	return rf
+}
+
+// NewResponseFilterer creates a new ResponseFilterer with the given parameters.
+// It is used to filter the response based on the rules and authz data.
+func NewResponseFilterer(restMapper meta.RESTMapper, input *rules.ResolveInput, filteredRules []*rules.RunnableRule, client v1.PermissionsServiceClient) (*StandardResponseFilterer, error) {
+	return &StandardResponseFilterer{
+		restMapper: restMapper,
+		input:      input,
+
+		filteredRules: filteredRules,
+		client:        client,
+
+		prefilteredStarted: false,
+		preFilterCompleted: make(chan prefilterResult, 1),
+	}, nil
+}
+
+// NewResponseFiltererForWatch creates a new ResponseFilterer specifically for watch requests.
+func NewResponseFiltererForWatch(restMapper meta.RESTMapper, input *rules.ResolveInput, foundWatchRule *rules.RunnableRule, watchClient v1.WatchServiceClient, checkClient v1.PermissionsServiceClient) (*WatchResponseFilterer, error) {
+	return &WatchResponseFilterer{
+		restMapper:  restMapper,
+		input:       input,
+		watchRule:   foundWatchRule,
+		checkClient: checkClient,
+		watchClient: watchClient,
+	}, nil
+}
+
+// StandardResponseFilterer is used to filter the response based on the rules and authz data.
+type StandardResponseFilterer struct {
+	restMapper meta.RESTMapper
+	input      *rules.ResolveInput
+
+	filteredRules []*rules.RunnableRule
+	client        v1.PermissionsServiceClient
+
+	prefilteredStarted bool
+	preFilterCompleted chan prefilterResult
+}
+
+// RunPreFilters runs the pre-filters for the request, if any. If not invoked before the response is filtered,
+// an error will be returned by the FilterResp method.
+func (rf *StandardResponseFilterer) RunPreFilters(req *http.Request) error {
+	if rf.prefilteredStarted {
```

Comment:
is this variable always read from the same goroutine?

Reply:
Yep